### ISSUE TITLE
Tame ProductsCacheIntegrityCheckerJob

### ISSUE THREAD

**Issue by sauloperez** [openfoodfoundation/openfoodnetwork#4235]
### What we should change and why (this is tech debt)

As you can see from the schedule and code below

https://github.com/openfoodfoundation/openfoodnetwork/blob/e3ce2a7486c91a449802b8d41415d18778694c67/config/schedule.rb#L15-L17

(notice however that we used to run this every hour until https://github.com/openfoodfoundation/openfoodnetwork/issues/4210)

https://github.com/openfoodfoundation/openfoodnetwork/blob/e3ce2a7486c91a449802b8d41415d18778694c67/lib/tasks/cache.rake#L5-L10

We periodically check if all the exchanges' cache is up-to-date which hits the server with a considerable (or even worse) load without solving the problem. It just tells we're missing cache refreshing logic somewhere.

#### What to do?

I suggest we:

* Spread the workload adding a much needed 2nd DelayedJob worker
* Refresh the cache instead of just sending a notification
* Send a monitoring metric to Datadog instead of polluting Bugsnag

Validate and reassess whether this is enough or we should:

* Spread the workload even more by processing them in batches during the night.

Eventually, we might need to rewrite the cache following a different strategy. See the proposal of [key-based cache expiration](https://github.com/openfoodfoundation/openfoodnetwork/issues/4039#issuecomment-511475614)

### Context

This stems from the zillions of Bugsnag errors that pile up in all accounts but for which we are taking no action.

We also realized how it affects CPU but yesterday became pretty clear that it was one of the culprits, if not the biggest one in https://github.com/openfoodfoundation/openfoodnetwork/issues/4210

### Impact and timeline

This jobs has proven to bring the UK's server to its limit and chances are that it does the same with others. Although we lack APM in Katuma, I'm quite sure the downtime experienced on the 2nd of September it was due to this as well.

Besides, all these suggestions are quite straightforward to implement as mean just readjusting what we already have.

**Comment by luisramos0** [comment]
looks good! anyone sees any issues with having 2 delayed job workers?

Now that we are running this job only once a day this is lower priority than before 🎉

**Comment by sauloperez** [comment]
Related to https://github.com/openfoodfoundation/openfoodnetwork/issues/3830

**Comment by Matt-Yorkley** [comment]
So this will increase the frequency of triggering the cache refresh job? That one's problematic as well in terms of resources. Maybe we need a better strategy for figuring out which parts of the app cause the cache to be out of sync?

**Comment by Matt-Yorkley** [comment]
Or just jump to the key-based cache option...

**Comment by mkllnk** [comment]
>     Spread the workload adding a much needed 2nd DelayedJob worker

Why do you think the second worker would solve the problem? The main issue on the UK server was overall server load. Two workers can increase the total load by doing more in parallel. We also need to be careful with jobs that may not be thread safe, e.g. the subscriptions jobs.

>     Refresh the cache instead of just sending a notification

Yes, I totally agree that it should fix the problem when it was detected. It's insane that we were doing all this computation work to then throw out the result.

>     Send a monitoring metric to Datadog instead of polluting Bugsnag

I still think that in invalid cache is a bug here. It shouldn't happen. Unfortunately, that logic has been buggy for years now.

**Comment by sauloperez** [comment]
> Why do you think the second worker would solve the problem? The main issue on the UK server was overall server load. Two workers can increase the total load by doing more in parallel. 

Totally right but now this happens at 01am. We can take the most out of the server by splitting the workload. Then, validate it with the metrics and if not remove the change `#agile`.

> We also need to be careful with jobs that may not be thread safe, e.g. the subscriptions jobs.

I'm not sure about that. AFAIK Delayed Job does not use threads just a process per worker thus, it's not a problem for us. It'd be if we used Sidekiq if I'm not mistaken. Am I wrong? I'm not much into this field.

> I still think that in invalid cache is a bug here. It shouldn't happen. Unfortunately, that logic has been buggy for years now.

Sure. I agree with you. That's why I want to be notified but it's useless if it doesn't get fixed. The same as not having it.

That being said, properly refreshing the cache is something we are looking into as part of the S2 we're working on. I think it was on @Matt-Yorkley 's plate.

**Comment by mkllnk** [comment]
>>    We also need to be careful with jobs that may not be thread safe, e.g. the subscriptions jobs.

> I'm not sure about that. AFAIK Delayed Job does not use threads just a process per worker thus, it's not a problem for us. It'd be if we used Sidekiq if I'm not mistaken. Am I wrong? I'm not much into this field.

Sorry, thread safe was the wrong term. I meant that the subscriptions jobs don't use locking and parallel jobs could mean that we place orders twice. We need to review all jobs to make sure that they are "parallel safe". Otherwise we risk some really difficult to track bugs.

Subscription example:
https://github.com/openfoodfoundation/openfoodnetwork/blob/908e1dfcafb7a4e7ce79f0f12a27305a62dbbae7/app/jobs/subscription_placement_job.rb#L5-L10

This is executed every five minutes and shouldn't be a problem, *unless* jobs are queued and then processed at once or something takes longer than five minutes.

**Comment by sauloperez** [comment]
Understand, we should check that for all jobs.

In this particular case, IMO `#initialize_order!` should take `placed_at` into account so we don't run into this issue no matter what.

Besides, various other solutions come to mind:
* Give more bandwidth by executing it every 15min. I don't think we need that "realtimeliness", tbh.
* Perform the individual order processing in a separate job together with something like https://github.com/bmpercy/unique_delayed_job. I bet this would also increase the server's concurrency since the processing of the proxy orders wouldn't block a worker for a long time as it does now.

**Comment by sauloperez** [comment]
Bulking a brain dump here. IMO it's time to revisit any products cache leftovers does something to your mind @Matt-Yorkley ? There are at least `touch`  calls we could remove.

**Comment by luisramos0** [comment]
ah, yeah, on enterprises. was that only for the cache?

**Comment by sauloperez** [comment]
the cached JSON was rather rich so I bet there are many models that logic related to it. We need to refresh our investigations and :fire: what might be left.

**Comment by Matt-Yorkley** [comment]
I removed as much as I could see, but there could be a few bits left.
